pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,396 @@ import (
+	for _, topology := range t.topologies {
+		if topology.IsOwnedBy(p.UID) {
+			topology.RemoveOwner(p.UID)
 		}
-		for _, pod := range topologyGroup.Pods {
-			domain := topologyGroup.NextDomain(constraints.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...).
-				Get(topologyGroup.Constraint.TopologyKey).
-				Values())
-			pod.Spec.NodeSelector = functional.UnionStringMaps(pod.Spec.NodeSelector, map[string]string{topologyGroup.Constraint.TopologyKey: domain})
+	}
+
+	// need to first ensure that we know all of the topology constraints.  This may require looking up
+	// existing pods in the running cluster to determine zonal topology skew.
+	if err := t.updateTopologySpread(ctx, p); err != nil {
+		errs = multierr.Append(errs, fmt.Errorf("tracking topology spread, %w", err))
+	}
+	if err := t.updateAffinity(ctx, p); err != nil {
+		errs = multierr.Append(errs, fmt.Errorf("tracking affinity topology, %w", err))
+	}
+	return errs
+}
+
+// updateInverseAffinities is used to identify pods with anti-affinity terms so we can track those topologies.  We
+// have to look at every pod in the cluster as there is no way to query for a pod with anti-affinity terms.
+func (t *Topology) updateInverseAffinities(ctx context.Context) error {
+	var nodeList v1.NodeList
+	if err := t.kubeClient.List(ctx, &nodeList); err != nil {
+		return fmt.Errorf("listing nodes, %w", err)
+	}
+	for i := range nodeList.Items {
+		var podlist v1.PodList
+		if err := t.kubeClient.List(ctx, &podlist, client.MatchingFields{"spec.nodeName": nodeList.Items[i].Name}); err != nil {
+			return fmt.Errorf("listing pods on node %s, %w", nodeList.Items[i].Name, err)
+		}
+		for j := range podlist.Items {
+			if pod.HasRequiredPodAntiAffinity(&podlist.Items[j]) {
+				if err := t.updateInverseAntiAffinity(ctx, &nodeList.Items[i], &podlist.Items[j]); err != nil {
+					return fmt.Errorf("tracking existing pod anti-affinity, %w", err)
+				}
+			}
 		}
 	}
 	return nil
 }
 
-// getTopologyGroups separates pods with equivalent topology rules
-func (t *Topology) getTopologyGroups(pods []*v1.Pod) []*TopologyGroup {
-	topologyGroupMap := map[uint64]*TopologyGroup{}
-	for _, pod := range pods {
-		for _, constraint := range pod.Spec.TopologySpreadConstraints {
-			// Add to existing group if exists, using a hash for efficient collision detection
-			key := topologyGroupKey(pod.Namespace, constraint)
-			if topologyGroup, ok := topologyGroupMap[key]; ok {
-				topologyGroup.Pods = append(topologyGroup.Pods, pod)
-			} else {
-				topologyGroupMap[key] = NewTopologyGroup(pod, constraint)
+func (t *Topology) updateTopologySpread(ctx context.Context, p *v1.Pod) error {
+	for _, cs := range p.Spec.TopologySpreadConstraints {
+		topologyGroup, err := t.newForSpread(p.Namespace, cs)
+		if err != nil {
+			return err
+		}
+		hash := topologyGroup.Hash()
+		if existing, ok := t.topologies[hash]; !ok {
+			if err := t.initializeTopologyGroup(ctx, topologyGroup); err != nil {
```

Comment:
Thoughts on moving initialization logic to inside `newForSpread` and then only new it after we check for existing?

Reply:
The Hash() method is on the result of the new call.  We do a cheap new, essentially just to see if it exists, and if not continue with the real init.

Reply:
Oh ofc. yes.